File: DataStructure/src/nonrecursiveQSort.dart
import 'dart:io';
import 'dart:convert';

int partition(List<int> list, int low, int high) {
  var mid = low + (high - low) ~/ 2;
  var temp = list[low];
  list[low] = list[mid];
  list[mid] = temp; // 枢轴元素交换到首位
  var j = low + 1; // j指向第一个大于枢轴元素的位置
  for (var i = low; i <= high; i++) {
    if (list[low] > list[i]) {
      // 比枢轴元素小的元素交换到前面
      temp = list[i];
      list[i] = list[j];
      list[j++] = temp;
    }
  }
  temp = list[low];
  list[low] = list[j - 1];
  list[j - 1] = temp; // 枢轴元素归位
  return j - 1;
}

/// 快速排序非递归算法实现
void nonrecursiveQSort(List<int> list) {
  List<int> stack = [];
  stack.addAll([0, list.length - 1]);
  while (stack.isNotEmpty) {
    // 栈不空
    var high = stack.removeLast();
    var low = stack.removeLast();
    var pivot = partition(list, low, high); // 一趟划分，返回枢轴下标索引
    if (pivot - 1 > low) {
      // 子序列长度大于1，入栈
      stack.addAll([low, pivot - 1]);
    }
    if (pivot + 1 < high) {
      // 子序列长度大于1，入栈
      stack.addAll([pivot + 1, high]);
    }
  }
}

void main() async {
  List<int> a = [];
  var s = (await stdin.transform(ASCII.decoder).toList())
      .join()
      .trim()
      .split(new RegExp(r'[\n\r\n\s+]+'));
  s.forEach((i) => a.add(int.parse(i)));
  nonrecursiveQSort(a);
  var buffer = new StringBuffer();
  a.forEach((i) => buffer.write(' $i'));
  print(buffer.toString().trimLeft());
}

File: DataStructure/src/nonrecursiveQSort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const STACK_SIZE = 1000 // 栈最大深度

func partition(list []int, low, high int) int {
	mid := low + (high-low)/2
	list[low], list[mid] = list[mid], list[low] // 枢轴元素交换到首位
	j := low + 1                                // j指向第一个大于枢轴元素的位置
	for i := low; i <= high; i++ {
		if list[low] > list[i] { // 比枢轴元素小的元素交换到前面
			list[i], list[j] = list[j], list[i]
			j++
		}
	}
	list[low], list[j-1] = list[j-1], list[low] // 枢轴元素归位
	return j - 1
}

// 快速排序非递归算法实现
func nonrecursiveQSort(list []int) {
	var stack [STACK_SIZE]int
	top := -1 //指向栈顶元素
	top++
	stack[top] = 0
	top++
	stack[top] = len(list) - 1
	for top > -1 { // 栈不空
		low, high := stack[top-1], stack[top] // 记录待排子序列起、始位置
		top -= 2                              // 出栈
		pivot := partition(list, low, high)   // 一趟划分，返回枢轴下标索引
		if pivot-1 > low {                    // 子序列长度大于1，入栈
			top++
			stack[top] = low
			top++
			stack[top] = pivot - 1
		}
		if pivot+1 < high { // 子序列长度大于1，入栈
			top++
			stack[top] = pivot + 1
			top++
			stack[top] = high
		}
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	scanner.Split(bufio.ScanWords)
	var a []int
	for scanner.Scan() {
		v, _ := strconv.Atoi(scanner.Text())
		a = append(a, v)
	}
	nonrecursiveQSort(a)
	for i, v := range a {
		if i == 0 {
			fmt.Fprint(w, v)
		} else {
			fmt.Fprint(w, " ", v)
		}
	}
	w.Flush()
	fmt.Println()
}
